woocommerce/woocommerce-gutenberg-products-block
# Allow block templates to be customised and saved

This PR focuses on allowing blocks templates to be customised, saved, and retrieved without errors.

The changes made are:
- Copy the following functions from Gutenberg (because they are private and we need access to them in Blocks) into our `BlockTemplateUtils` class.
  - `gutenberg_add_template_info`
  -  `gutenberg_build_template_result_from_post`
- Wherever a `$template`'s properties are accessed, normalise `$template` to be an object. This is necessary because this is sometimes an array or sometimes a `WP_Block_Template`.
- When building templates from the woo-blocks filesystem, ensure the theme is set to `woocommerce`. This is necessary as the current behaviour is to save the template under the current theme's slug, which seems incorrect.
- Add a `maybe_return_blocks_template` function - this is used to short-circuit the `gutenberg_get_block_templates` method from Gutenberg. It checks the db to see if there's a saved template already. If there is then we can return that.
- Add `get_single_block_template` function. Until now there was no way to get a single woo-blocks block template in our codebase. This function is used within `gutenberg_get_block_templates` to override/augment the data returned by the `gutenberg_get_block_templates` function. This is how we inject _our_ templates into the list of templates gutenberg finds.
- Update `get_block_templates` in `BlockTemplatesController.php` to check the database first to see if any saved templates from the `woocommerce` theme exist. If they do, build a template from the saved posts.
- Update `add_block_templates` in `BlockTemplatesController.php` to only add the templates from files that don't have a customised version in the database.
- If the query passed to the `get_block_templates` filter contains `slug__in` then filter the results of `BlockTemplatesController`'s `get_block_templates` method to only included those that match the queried slugs.
- Add a `remove_theme_templates_with_custom_alternative` - this will cater for the situation described in *Second testing steps for alternative scenario (Theme adds a template for an already customised template)*
 
Fixes #5024 




#### Other Checks


### Screenshots



### Testing

#### Automated Tests
### Manual Testing

How to test the changes in this Pull Request:
1. Use an FSE theme.
2. Add a file named `single-product.html` to `woocommerce-gutenberg-products-block/templates/block-templates/` - insert the following content:
```

 <p>single-product</p> 


```
3. On the site editor change the content of this file, add a block, move things about, etc. 
4. Open the network tab in the inspector and save the page. Ensure there are no errors.
5. Reload the editor, verify your changes have persisted.
6. Delete the `single-product.html` file 
7. Reload the editor, verify your changes have persisted.
8. Go to `Appearance > Templates` and verify the template is saved with a theme of `woocommerce`.
9. Delete the template.
10. Add `single-product.html` with the above contents to `wp-content/themes/<your-theme>/block-templates/` and repeat steps 3-6. 

### Second testing steps for alternative scenario (Theme adds a template for an already customised template)
1. Delete all saved templates from the Appearance > Templates page.
2. Add the `single-product.html` file above to `woo-blocks/templates/block-templates`. Ensure there is *not* a `single-product.html` file in your theme's block-templates directory!
3. Go to the site editor, and verify you can edit the template.
4. Customise it and save it.
5. Add the `single-product.html` file to your theme's block-templates directory. 
6. Reload the editor and ensure your customised version of `single-product` is loaded. 
7. Ensure there is only one entry for the template in the `General templates` list in the editor.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).

1. Install a FSE theme, i.e. TT1 Blocks, or Tove.
2. Make make sure you have [this change applied in your WooCommerce plugin](https://github.com/woocommerce/woocommerce/pull/30997/files).
The block theme needs to declare `add_theme_support( 'woocommerce' );` in the `functions.php` file.
3. Go to the site editor and use the arrow at the top of the page to select a template. Select General Templates > Single product.
4. Modify the template, add some blocks, move some things around etc.
5. Save the template.
6. Reload the page, verify the changes have persisted.
7. Go to the dashboard then `Appearance > Templates`. Verify there's a template called `Single product` with the theme `woocommerce`
